kernelng.8
.TH "kernelng" "8" "0.1.0"

.SH "NAME"
.LP
kernelng \- A tool to create and manage overlays of ebuilds based on kernel-ng.eclass

.SH "SYNTAX"
.LP
kernelng [options] [sub-command [sub-command options]]

.SH "OPTIONS"
.TP
.B \-h, \-\-help
Print a usage summary and exit.
.TP
.B \-\-version
Show program's version number and exit.
.SS "Main modes"
.TP
.B \-f, \-\-foo
This will jack shit up and make it all snazzy and
shit.  You know.
.TP
.B \-b, \-\-bar
Sometimes you need a drink and you don't wanna make it yourself.
In bar mode, kernelng will provide you a comfortable seat and
make your drinks for you for a moderate fee.
.TP
.BI \-R " REGION " "\fR,\fP \-\-region " REGION "
Only use mirrors from the specified region.
NOTE: Names with a space must be quoted.
    eg.:  -R 'North America'
.SS "Other options"
.TP
.BI \-b " BLOCKSIZE " "\fR,\fP \-\-blocksize" " BLOCKSIZE "
This is to be used in automatic mode and will split the hosts into blocks of
BLOCKSIZE for use with netselect. This is required for certain routers which
block 40+ requests at any given time.
Recommended parameters to pass are: -s 3 -b 10
.TP
.B \-o, \-\-output
Output Only Mode, this is especially useful when being used during installation,
to redirect output to a file other than /etc/portage/make.conf.
.TP
.B \-q, \-\-quiet
Quiet mode.

.SH "EXAMPLES"
automatic:
.LP
.RS 8
Autoselect the 5 best mirrors and save results in /etc/portage/make.conf,
a backup will be created.
.LP
# mirrorselect -s5
.LP
Autoselect the 3 best mirrors, use a
.I BLOCKSIZE
of 10 and append the output to /etc/portage/make.conf.
.LP
# mirrorselect -s3 -b10 -o >> /etc/portage/make.conf
.LP
Autoselect the 4 best mirrors, it also loads a 100k file from each mirror.
The result will be written to /etc/portage/make.conf, a backup will be created.
.LP
# mirrorselect -D -s4
.LP
.RE
interactive:
.LP
.RS 8
Start mirrorselect in interactive mode to select rsync mirrors.
.LP
# mirrorselect -i -r
.RE

.SH "AUTHORS"
.LP
Colin Kingsley <tercel@gentoo.org>
.LP
Zac Medico <zmedico@gentoo.org>
.LP
Sebastian Pipping <sebastian@pipping.org>
.LP
Christian Ruppert <idl0r@gentoo.org>
.LP
Brian Dolbec <dolsen@gentoo.org>
.LP
Douglas Freed <dwfreed@mtu.edu>
.LP
Gregory Turner <gmt@be-evil.net>

.SH "BUGS"
Please report any bugs to https://github.com/gmt/kernel-ng-util/issues


kernelng/scripts/kernelng.py
#!/usr/bin/env python
#-*- coding:utf-8 -*-
# vim:ai:sta:et:ts=4:sw=4:sts=4

"""kernelng 0.x
 Tool for maintaining customized overlays of kernel-ng.eclass-based ebuilds

Copyright 2005-2014 Gentoo Foundation

        Copyright (C) 2005 Colin Kingsley <tercel@gentoo.org>
        Copyright (C) 2008 Zac Medico <zmedico@gentoo.org>
        Copyright (C) 2009 Sebastian Pipping <sebastian@pipping.org>
        Copyright (C) 2009 Christian Ruppert <idl0r@gentoo.org>
        Copyright (C) 2012 Brian Dolbec <dolsen@gentoo.org>
        Copyright (C) 2014 Gregory M. Turner <gmt@be-evil.net>

Distributed under the terms of the GNU General Public License v2
 This program is free software; you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, version 2 of the License.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA.

"""

from __future__ import print_function

import sys
import os
import re

import click
from kernelng.kngclick import kngcommand, knggroup, OCTAL_3

import portage

# This block ensures that ^C interrupts are handled quietly.
try:
    import signal

    def exithandler(signum,frame):
        signal.signal(signal.SIGINT, signal.SIG_IGN)
        signal.signal(signal.SIGTERM, signal.SIG_IGN)
        click.echo('Caught signal %s. Exiting' % signum)
        sys.exit(1)

    signal.signal(signal.SIGINT, exithandler)
    signal.signal(signal.SIGTERM, exithandler)
    signal.signal(signal.SIGPIPE, signal.SIG_DFL)

    # eprefix compatibility
    try:
        from portage.const import rootuid
    except ImportError:
        rootuid = 0

    # establish the eprefix, initially set so eprefixify can
    # set it on install
    EPREFIX = "@GENTOO_PORTAGE_EPREFIX@"

    # check and set it if it wasn't
    if EPREFIX == "@GENTOO_%s_EPREFIX@" % "PORTAGE":
        EPREFIX = ''

    PROGNAME = sys.argv[0].split(os.path.sep)[-1] if len(sys.argv) >= 1 else 'kernelng'
    PROGDESC = 'kernel-ng-util'
    FRAMEWORK = 'kernel-ng'
    SUBCMDHELP = 'For detailed help with subcommands, issue the sub-command' \
        ' followed by the %(helpshort)s or %(helplong)s option.' % {
            'helpshort': click.style('-h', fg='white', bold=True),
            'helplong': click.style('--help', fg='white', bold=True),
        }
    HS_RE = re.compile('%\([^)]*\)[^\W\d_]', re.UNICODE)
    HS = {
        'prog': click.style(PROGNAME, fg='white', bold=True),
        'progdesc': PROGDESC,
        'framework': FRAMEWORK,
        'eprefix': EPREFIX,
        'subcmdhelp': SUBCMDHELP,
    }
    def hs(text):
        return text % HS if re.search(HS_RE, text) else text

    CONTEXT_SETTINGS = dict(
        help_option_names = ['-h', '--help']
    )

    # python3 wigs out if these are proxies
    portage_uid = int(portage.data.portage_uid)
    portage_gid = int(portage.data.portage_gid)

    @knggroup(
        PROGNAME,
        context_settings=CONTEXT_SETTINGS,
        help = hs(
            """
            %(progdesc)s provides the %(prog)s command to manage
            a site-specific overlay containing customized %(framework)s
            packages.  The following example sequence
            of commands could be used to configure and deploy
            a %(framework)s package from scratch:

            \b
              $ sudo %(prog)s config -i
              $ sudo %(prog)s overlay create
              $ sudo %(prog)s ebuild create ng
              $ sudo emerge -u @world

            %(subcmdhelp)s
            """
        )
    )
    def cli():
        pass

    @cli.knggroup(
        help = hs(
            """
            Manage the %(progdesc)s overlay.

            %(subcmdhelp)s
            """
        )
    )
    def overlay():
        pass

    @overlay.kngcommand(
        help = hs(
            """
            Creates and activates an empty %(progdesc)s overlay.
            """
        ),
        short_help = hs("Create and activate an empty %(progdesc)s overlay.")
    )
    @click.option('-u', '--uid', type=click.INT, default=portage_uid,
        help='Numeric user id to assign to overlay files.')
    @click.option('-g', '--gid', type=click.INT, default=portage_gid,
        help='Numeric group id to assign to overlay files.')
    @click.option('-p', '--perm', type=OCTAL_3, default=0o664,
        help='Three-digit octal permissions to assign to overlay files.')
    def create(uid, gid, perm):
        click.echo('uid=%d gid=%d perm=0%o' % (uid, gid, perm))

    @overlay.kngcommand(
        help = hs(
            """
            Deactivates and/or removes the %(progdesc)s overlay.
            """
        ),
        short_help = hs("Deactivate or remove %(progdesc)s overlay.")
    )
    def destroy():
        pass

    @cli.knggroup(
        help = hs(
            """
            Modify the %(progdesc)s configuration.

            %(subcmdhelp)s
            """
        )
    )
    def config():
        pass

    if __name__ == '__main__':
        cli()

except KeyboardInterrupt:
    click.echo('Exited due to keyboard interrupt')
    sys.exit(130)
